### ISSUE TITLE
Reported total_problems seems inconsistent when using multiple nodes

### ISSUE THREAD

**Issue by faustomilletari** [PrimeIntellect-ai/prime-rl#956]
Hello,

I was looking to quantify the speedup brought by using multiple nodes.

When I look at total_problems vs. time I see this

<img width="1055" height="526" alt="Image" src="https://github.com/user-attachments/assets/980256d6-c291-448a-86b6-9c79d07cd0ef" />

It looks like that no matter the number of nodes, the speedup is negligible. 

I have looked also at time per step

<img width="1055" height="526" alt="Image" src="https://github.com/user-attachments/assets/c6f39673-6722-4588-bb4f-9e2d396f2500" />

it looks like each node takes more or less the same time per step, but that should also mean that the total_problem should increase almost linearly with the number of nodes. To my understanding if I have for example micro_batch_size 1 and number_rollouts 8 and I have 1 node with 8 GPUs, each step will use all GPUs to compute 8 problems. Having the same setting with 4 nodes with 8 GPUs each, one step yields the computation of 32 problems. 

I was therefore expecting to see a noticeable increase in the slope of the plots of total_problems vs. time when using multiple nodes. 
 
(this of course assuming that I have correctly interpreted the meaning of total_problems)

**Comment by mikasenghaas** [comment]
hey @faustomilletari! it would be really helpful if you could give some more details on the setup here:
- which model are you using?
- what type of node are you using?
- i am assuming you are running a full training on 1,2,4 nodes respectively? if so, how many gpus are you using for trainer and inference? what parallelism strategy do you use on inference (e.g. dp/tp) ?
- what other configs are you using, e.g. batch size, seq len, rollouts per example?

if you do not vary batch size and rollouts per example, then it is indeed not surprising that the total number of problems is the same independently of how many nodes you run it on. at each step you are getting rollouts for `num_examples // rollouts_per_example` examples. it doesn't matter how many gpus your inference server uses.

the fact that your step time is not decreasing means that the additional nodes do not yield any performance gain which is ofc undesirable. it is hard to say from afar why that might be, but my guess is that if you train relatively small model and/or use low seq len/ batch size which makes scaling out the inference unncessary

**Comment by faustomilletari** [comment]
Hello,

please find here some details about my training job.

1. Qwen/Qwen2.5-3b-Instruct

2. H100 / 8GPUs

3. 1 node training + 1 inference; 2 nodes training + 1 inference; 4 nodes training + 1 inference

4. Parameter:
- max_steps = 100000
- batch_size = 256
- micro_batch_size = 1
- seq_len = 4096
- rollouts_per_example = 16
- mask_truncated_completions = true

the model is not very big, but I was expecting a speedup as more micro-batches could be processed at the same time on multiple nodes. 

**to my understanding** the step number means how many parallel micro batches have been processed. 
and in this case all jobs have more or less the same time per step. The 1 Node job will process 8 micro batches per step, the 2 Nodes job is going to process 16 micro batches and the 4 Nodes job is going to process 32 micro batches per step.

it appears that across jobs steps take the same time.

<img width="1092" height="526" alt="Image" src="https://github.com/user-attachments/assets/8a874475-a4df-4c48-b2f3-f45b9c5267e0" />

 I was expecting the total number of problems and total tokens to be 1x for 1 Node job, almost 2x for 2 Nodes job and almost 4x for 4 Nodes job.


I also see this, 

<img width="1092" height="526" alt="Image" src="https://github.com/user-attachments/assets/77205144-dd7c-4c8b-8fad-3713f28db42a" />

which is kind of weird if you notice in the legend which run take how much time.

**Comment by mikasenghaas** [comment]
ah, thanks for sharing, i think i know what is happening now. it seems like you are always using the same number of gpus (1) for inference and only scaling up the number of trainer nodes. it is very likely that your inference is bottlenecking your trainer. to check this you can look at the `time/wait_for_weight_ckpt` which is the time the orchestrator waits to receive the updated policy and `time/wait_for_batch` which is the time the trainer waits for the rollouts to train on. i'd bet that that `time/wait_for_batch` is pretty high for your runs.

<img width="653" height="239" alt="Image" src="https://github.com/user-attachments/assets/72f198bc-50ba-408b-accf-ab4aef32cdc5" />

to understand what's happening take this simplified scenario of one-step off policy (async level=1) training:
1. step 0: inference is generating rollouts 0 in time x
2. step 1: inference is generating rollouts 1 in time x, trainer is training on rollouts from step 0 in time y
3. step 2: inference is generating rollouts 2 in time x, trainer is training on rollouts from step 1 in time y

maybe this graphic explains it better than my words:)

<img width="794" height="125" alt="Image" src="https://github.com/user-attachments/assets/c77c9b4a-67ea-4389-8da7-6e11cff6def7" />

note that at each step the trainer cannot start working before inference at the previous step is done. likewise inference cannot start working without the updated weights from the trainer in the previous step. now if x=y (which should generally be your aim in your hardware choice), you suffer on idle time and train optimally. if x<y, then your trainer is bottlenecking the training and you should increase the number of trainer gpus. if x>y, then inference is bottlenecking your training and you should use more inference gpus. 

typically, inference is the bottlenecking component, although this ofc depends on the batch size, seq len, avg model response length and so forth and actually should be pretty carefully benchmarked before deciding on a hardware configuration. typically we train with ratios 1:1, 1:2 or 1:3 of trainer:inference gpus as a general guidance.

**Comment by mikasenghaas** [comment]
> to my understanding the step number means how many parallel micro batches have been processed.
and in this case all jobs have more or less the same time per step. The 1 Node job will process 8 micro batches per step, the 2 Nodes job is going to process 16 micro batches and the 4 Nodes job is going to process 32 micro batches per step.

i might be misunderstanding smth here but typically by step we mean a "training step" in the sense that one optimizer step has been performed (the weights of your model have been updated once). it doesn't really matter in this definition in how many micro batches your full training batch is split.

in general, scaling up your node count doesn't magically scale the batch size or number of micro batches. they are always configured globally. for example, say you have `batch_size=256` and `micro_batch_size=1`, then there are 256 micro batches.
- if you have one gpu, then it will process all 256 micro batches
- if you have two gpus, each will process 128 micro batches
- and so forth

hope this can clarify some things!

**Comment by faustomilletari** [comment]
Hello, thanks for your thorough answers.

The time/wait_for_batch seems pretty similar across runs. **I use 8 H100 for inference**. I believe that using 8 H100 via VLLM (which is notoriously very efficient) may be enough to feed 4 training nodes (32 GPUs total).

<img width="1009" height="526" alt="Image" src="https://github.com/user-attachments/assets/320bd6c6-a657-4407-b8d3-caefc4a8031c" />

<img width="1009" height="526" alt="Image" src="https://github.com/user-attachments/assets/cbc08e62-881f-4bd8-aecc-fe7b4a44615a" />

What I meant by "step" is the **step number**, as it appears in the horizontal axis of the weights and biases graphs. To my understanding 1 training step == 1 micro batch fully processed (fwd and back).

1 training step for me is whatever produces this log:

```
[19:19:04][💡][model_utils:log_w...] 19:18:31    INFO Starting training step 702
[19:19:04][💡][model_utils:log_w...] 19:18:31    INFO Waiting for training batch to arrive
[19:19:04][💡][model_utils:log_w...] 19:18:53    INFO Starting forward and backward pass (num_micro_batches=1)
[19:19:04][💡][model_utils:log_w...] 19:18:53 SUCCESS Step 702 | Time: 22.84s | Loss: -0.0000 | Entropy: 0.1259 | Importance Ratio: 0.9823 | Grad. Norm: 0.0039 | LR: 1.00e-06 | Throughput: 4932 tokens/s | MFU: 0.3%
```

which is reported in weights and biases in correspondence of x=702.

If I go and access the logs of each Node, **each of them is doing step 702 ** independently. In my runs with 1, 2 and 4 training nodes (and 1 inference node) I see that the time it takes to do one step (plotting #step vs. time) is basically the same across training runs and nodes (note that horizontal axis is now "relative time"). 

<img width="797" height="417" alt="Image" src="https://github.com/user-attachments/assets/23208ac8-f8ed-479e-acf7-8e87e6b50b37" />

this means that the job using 1 training node (8 GPUs) is doing 8/256th of a batch at each training step. The node using 4 nodes (32 GPUs) is doing 32/256th of a batch at each training step reported in wandb. They take the same time to do so: this means that the 4 nodes job is roughly 4x faster than the 1 node job.

I don't see this reflected in progress/total_problems

<img width="825" height="487" alt="Image" src="https://github.com/user-attachments/assets/5ed8737b-26df-4e3d-b60c-df01d386db02" />

How can it be that 4 Nodes doing 8 micro batches **each** and taking time T for each step progress at the same pace as a job running a single node and taking the same time T for each step?

**Comment by faustomilletari** [comment]
Upon reflection I understood that the root cause of my problems is that communication overhead is a couple of orders of magnitude slower than the fwd/bckward step computation in my case, and one order of magnitude slower than inference time for completions. 

This is absolutely killing my performance. I'll try to find a solution and report back.
